### PR TITLE
Makeflow Amazon Lambda Support

### PR DESCRIPTION
**Pull request by dthain**
Merged Conrad's branch with master, making a new PR to fix up a few issues for consistency.

### CODE REVIEW

**Comment by dthain** [comment]
Supersedes PR #1560

**Comment by dthain** [comment]
@nhazekam @Nekel-Seyew please take a look and see if anything is amiss.

File: makeflow/src/makeflow_lambda_setup [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	exit 1
+fi
+
+# Create the role that will be attributed to the Lambda function. This
+# gives the Lambda function 'permission' to access S3 and to execute
+
+echo -n "Creating role policy file $ROLE_POLICY_FILE..."
+
+# The skeleton of the role, to which we will attach the policies
+cat > "$ROLE_POLICY_FILE" <<- EOF
```
Comment (Nekel-Seyew) [inline makeflow/src/makeflow_lambda_setup]: The user needs permissions to create the roll and add it to themselves. Additionally, should the user already have these privileges before they can access the tools? That way if it's a student, the professor knows who has what permissions, and we're not adding powers to the users that they're not cleared to have.

Comment (dthain) [reply makeflow/src/makeflow_lambda_setup]: AFAIK, no special privileges are required b/c this is creating a fresh role unrelated to exisiting roles.  The user is simply granting to this function the ability to access this particular S3 bucket, so it's all self contained, I think.

Maybe see if you can run it yourself and run into any problems.

**Comment by dthain** [comment]
@nkremerh we need someone with a fresh account that can try this out from a new user perspective.
Please check out this branch, read the (brief) docs about Amazon Lambda support, set up an AWS account, and then see if you are able to get a simple makeflow to run by following the instructions.

File: makeflow/src/makeflow_lambda_setup (hunk 1 above)
Comment (dthain) [reply makeflow/src/makeflow_lambda_setup]: Ah, I didn't read Conrad's code closely enough.  This creates a policy that let's the function access anything in S3 (that the user can access).  It should instead just grant access to a single bucket, e.g.:

https://aws.amazon.com/blogs/security/writing-iam-policies-how-to-grant-access-to-an-amazon-s3-bucket/

**Comment by dthain** [comment]
@nkremerh please try again and see if this solves your python issue.

**Comment by dthain** [comment]
@Nekel-Seyew take a look at how I downgraded from json to nvpair in the setup/cleanup/batch modules.  This way, the user doesn't require a specific version of python to make things work.  (@nkremerh crashed due to the usual python 2/3 conflicts.)  You probably want to do this in Amazon Batch as well.

**Comment by nkremerh** [comment]
@dthain Ran BWA again with the changes, worked perfectly.

**Comment by Nekel-Seyew** [comment]
Ah, didn't think about that. The json module normally works pretty well. Ok, will look into changing that.

**Pull request merged**